dubbo-registry/dubbo-registry-redis/src/main/java/org/apache/dubbo/registry/redis/RedisRegistry.java

```diff
@@ -199,46 +164,43 @@ private void deferExpired() {
-                if (values != null && values.size() > 0) {
-                    boolean delete = false;
-                    long now = System.currentTimeMillis();
-                    for (Map.Entry<String, String> entry : values.entrySet()) {
-                        URL url = URL.valueOf(entry.getKey());
-                        if (url.getParameter(Constants.DYNAMIC_KEY, true)) {
-                            long expire = Long.parseLong(entry.getValue());
-                            if (expire < now) {
-                                jedis.hdel(key, entry.getKey());
-                                delete = true;
-                                if (logger.isWarnEnabled()) {
-                                    logger.warn("Delete expired key: " + key + " -> value: " + entry.getKey() + ", expire: " + new Date(expire) + ", now: " + new Date(now));
-                                }
-                            }
+        if (keys == null || keys.isEmpty()) {
+            return;
+        }
+        for (String key : keys) {
+            Map<String, String> values = jedis.hgetAll(key);
+            if (values == null || values.isEmpty()) {
+                continue;
+            }
+            boolean delete = false;
+            long now = System.currentTimeMillis();
+            for (Map.Entry<String, String> entry : values.entrySet()) {
+                URL url = URL.valueOf(entry.getKey());
+                if (url.getParameter(Constants.DYNAMIC_KEY, true)) {
+                    long expire = Long.parseLong(entry.getValue());
+                    if (expire < now) {
+                        jedis.hdel(key, entry.getKey());
+                        delete = true;
+                        if (logger.isWarnEnabled()) {
+                            logger.warn("Delete expired key: " + key + " -> value: " + entry.getKey() + ", expire: " + new Date(expire) + ", now: " + new Date(now));
                         }
                     }
-                    if (delete) {
-                        jedis.publish(key, Constants.UNREGISTER);
-                    }
                 }
             }
+            if (delete) {
+                jedis.publish(key, Constants.UNREGISTER);
+            }
         }
     }
 
     @Override
     public boolean isAvailable() {
         for (JedisPool jedisPool : jedisPools.values()) {
-            try {
-                Jedis jedis = jedisPool.getResource();
-                try {
-                    if (jedis.isConnected()) {
-                        return true; // At least one single machine is available.
-                    }
-                } finally {
-                    jedis.close();
+            try (Jedis jedis = jedisPool.getResource()) {
+                if (jedis.isConnected()) {
+                    return true; // At least one single machine is available.
                 }
-            } catch (Throwable t) {
+            } catch (Throwable ignored) {
             }

```

Comment:
Please close `jedis`  Calling the Close method will return the connection object to the object pool

Reply:
this type of try block will call the close method automatically

Reply:
Sorry, I don’t see it clearly, you are right.